SSA/Projects/InstCombine/ForLean.lean

```diff
@@ -3,6 +3,257 @@ import SSA.Projects.InstCombine.LLVM.Semantics
 import Mathlib.Tactic.Ring
 import Mathlib.Data.BitVec
 
+#check BitVec.toInt_eq_toNat_of_msb
+
+theorem toInt_lt {w : Nat} (x : BitVec w) : x.toInt < 2 ^ (w - 1) := by
+  simp only [BitVec.toInt, Nat.cast_pow]
+  by_cases hw : w = 0
+  · subst hw
+    simp [BitVec.eq_nil x]
+  · rw [←Nat.two_pow_pred_add_two_pow_pred (by omega), ←Nat.two_mul]
+    split
+    case neg.isTrue h =>
+      norm_cast
+      omega
+    case neg.isFalse h =>
+      rw [sub_lt_iff_lt_add]
+      norm_cast
+      omega
+
+theorem le_toInt {w : Nat} (x : BitVec w) : -2 ^ (w - 1) ≤ x.toInt := by
+  simp only [BitVec.toInt, Nat.cast_pow]
+  by_cases hw : w = 0
+  · subst hw
+    simp [BitVec.eq_nil x]
+  · rw [←Nat.two_pow_pred_add_two_pow_pred (by omega), ←Nat.two_mul]
+    simp only [zero_lt_two, mul_lt_mul_left, Nat.cast_ofNat]
+    split
+    case neg.isTrue h =>
+      norm_cast
+      omega
+    case neg.isFalse h =>
+      simp only [neg_le_sub_iff_le_add]
+      norm_cast
+      rw [←Nat.two_pow_pred_add_two_pow_pred (by omega), ←Nat.two_mul]
+      omega
+
+theorem toInd_add_toInt_lt_two_pow (x y : BitVec w) :
+    (x.toInt + y.toInt) < 2 ^ w := by
+  by_cases hw : w = 0
+  · subst hw
+    simp [BitVec.eq_nil x, BitVec.eq_nil y]
+  · norm_cast
+    rw [←Nat.two_pow_pred_add_two_pow_pred (by omega)]
+    have hx := toInt_lt x
+    have hy := toInt_lt y
+    push_cast
+    omega
+
+theorem bmod_pos_iff_of_pos_lt {x : Int} {y : Nat} (h1 : 0 ≤ x) (h2 : x < y) :
+    0 ≤ (x.bmod y) ↔ x < (y + 1) / 2 := by
+  simp [Int.bmod_def]
+  rw [Int.emod_eq_of_lt (by omega) (by omega)]
+  split
+  · omega
+  · omega
+
+theorem bmod_neg_iff_of_pos_lt {x : Int} {y : Nat} (h1 : 0 ≤ x) (h2 : x < y) :
+    (x.bmod y) < 0 ↔ (y + 1) / 2 ≤ x := by
+  simp [Int.bmod_def]
+  rw [Int.emod_eq_of_lt (by omega) (by omega)]
+  split
+  · omega
+  · omega
+
+@[simp]
+theorem toInt_add_mod_elim {x y : BitVec w} :
+    (x.toInt + y.toInt) % 2 ^ w = x.toInt + y.toInt := by
+  have ltx := toInt_lt x
+  have lty := toInt_lt y
+  have h : x.toInt + y.toInt < 2 * 2 ^ (w - 1) := by
+    sorry
+
+  have h : x.toInt + y.toInt < 2 ^ w := by
+    sorry
+  clear ltx lty
+  --rw [Int.add_comm, Nat.two_pow_pred_mul_two] at h
+  omega
+
+theorem uadd_overflow_eq {w : Nat} (x y : BitVec w) :
+    uadd_overflow x y = BitVec.carry w x y false := by
+  simp only [uadd_overflow, BitVec.carry]
+  by_cases h : 2 ^ w ≤ x.toNat + y.toNat <;> simp [h]
+
+theorem aaaa : (2 ^ w + 1) / 2 = 2 ^ (w - 1) := by
```

Comment:
```suggestion
theorem aaaa : (2 ^ w + 1) / 2 = 2 ^ w := by
```

That should be easier to prove

Reply:
I don't think the new theorem is true, no?

Reply:
Ah, I'd read it as `2 ^ (w + 1)`, in which case it would've been true; ignore my suggestion!